napari/utils/_proxies.py

```diff
@@ -0,0 +1,81 @@
+import re
+import warnings
+from typing import Any, Callable, Generic, TypeVar
+
+import wrapt
+
+from ..utils.translations import trans
+
+_T = TypeVar("_T")
+
+
+class ReadOnlyWrapper(wrapt.ObjectProxy):
+    """
+    Disable item and attribute setting with the exception of  ``__wrapped__``.
+    """
+
+    def __setattr__(self, name, val):
+        if name != '__wrapped__':
+            raise TypeError(
+                trans._(
+                    'cannot set attribute {name}',
+                    deferred=True,
+                    name=name,
+                )
+            )
+
+        super().__setattr__(name, val)
+
+    def __setitem__(self, name, val):
+        raise TypeError(
+            trans._('cannot set item {name}', deferred=True, name=name)
+        )
+
+
+_SUNDER = re.compile('^_[^_]')
+
+
+class PublicOnlyProxy(wrapt.ObjectProxy, Generic[_T]):
+    """Proxy to prevent private attribute and item access, recursively."""
+
+    __wrapped__: _T
+
+    def __getattr__(self, name: str):
+        if name.startswith('_'):
+            warnings.warn(
+                trans._(
+                    "Private attribute access in this context is deprecated and will be unavailable in version 0.4.14",
+                    deferred=True,
+                ),
+                category=FutureWarning,
+                stacklevel=2,
+            )
+            # name = f'{type(self.__wrapped__).__name__}.{name}'
+            # raise AttributeError(
+            #     trans._(
+            #         "Private attribute access ('{name}') not allowed in this context.",
+            #         deferred=True,
+            #         name=name,
+            #     )
+            # )
```

Comment:
should be removed?
```suggestion
```

Reply:
I left that in so that when we remove the deprecation warning you just need to uncomment...

Reply:
ah ha! thanks